pkg/cloud/azure/services/availabilitysets/availabilitysets.go

```diff
@@ -23,27 +24,45 @@ func (s *Service) CreateOrUpdate(ctx context.Context, spec azure.Spec) error {
 	if !ok {
 		return errors.New("invalid availability set specification")
 	}
-	faultDomainCount, err := s.getMaximumFaultDomainCount(ctx)
-	if err != nil {
-		return fmt.Errorf("failed to get fault domain count: %w", err)
+
+	existingAvailabilitySet, err := s.Client.Get(ctx, s.Scope.MachineConfig.ResourceGroup, availabilitysetsSpec.Name)
+	if err != nil && !azure.ResourceNotFound(err) {
+		return fmt.Errorf("failed to get availability set %s: %w", availabilitysetsSpec.Name, err)
 	}
 
-	asParams := compute.AvailabilitySet{
-		Name: to.StringPtr(availabilitysetsSpec.Name),
-		Sku: &compute.Sku{
-			Name: to.StringPtr(string(compute.AvailabilitySetSkuTypesAligned)),
-		},
-		Location: to.StringPtr(s.Scope.Location()),
-		AvailabilitySetProperties: &compute.AvailabilitySetProperties{
-			PlatformFaultDomainCount:  to.Int32Ptr(int32(faultDomainCount)),
-			PlatformUpdateDomainCount: to.Int32Ptr(int32(5)),
-		},
-		Tags: s.Scope.Tags,
+	var availabilitySet compute.AvailabilitySet
+	if azure.ResourceNotFound(err) {
+		faultDomainCount, err := s.getMaximumFaultDomainCount(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to get fault domain count: %w", err)
+		}
+
+		availabilitySet = compute.AvailabilitySet{
+			Name: to.StringPtr(availabilitysetsSpec.Name),
+			Sku: &compute.Sku{
+				Name: to.StringPtr(string(compute.AvailabilitySetSkuTypesAligned)),
+			},
+			Location: to.StringPtr(s.Scope.Location()),
+			AvailabilitySetProperties: &compute.AvailabilitySetProperties{
+				PlatformFaultDomainCount:  to.Int32Ptr(int32(faultDomainCount)),
+				PlatformUpdateDomainCount: to.Int32Ptr(int32(5)),
+			},
+			Tags: s.Scope.Tags,
+		}
+	} else {
+		if reflect.DeepEqual(existingAvailabilitySet.Tags, s.Scope.Tags) {
+			// Availability set already exists has desired tags.
+			// Other properties are immutable.
+			return nil
+		}
+
+		availabilitySet = existingAvailabilitySet
+		existingAvailabilitySet.Tags = s.Scope.Tags
```

Comment:
Did we support mutability of Tags this way before (I mean before we accidentally regressed it)? If not, I'd be inclined to leave this out. Maybe leave a comment that we *could* support mutable tags if we really wanted to.

If we leave it in, though, we should at least unit test it.

Reply:
`Did we support mutability of Tags this way before (I mean before we accidentally regressed it)?` 
We don't support it, the [enhancement](https://github.com/openshift/enhancements/blob/master/enhancements/api-review/azure_user_defined_tags.md) says it is out of scope, but it is implemented for every resource to update the tags.

This looks difficult to test. I think QE verify should be enough to test it still works.

The update of tags will break once we move to CAPZ anyway, as CAPZ does not update the availability set.